cryptoki/src/mechanism/kbkdf.rs

```diff
@@ -0,0 +1,371 @@
+                    ulValueLen: size_of::<CK_SP800_108_COUNTER_FORMAT>() as CK_ULONG,
+                },
+                PrfDataParamType::Counter(counter_format) => CK_PRF_DATA_PARAM {
+                    type_: CK_SP800_108_COUNTER,
+                    pValue: counter_format as *const _ as *mut _,
+                    ulValueLen: size_of::<CK_SP800_108_COUNTER_FORMAT>() as CK_ULONG,
+                },
+                PrfDataParamType::DkmLength(dkm_length_format) => CK_PRF_DATA_PARAM {
+                    type_: CK_SP800_108_DKM_LENGTH,
+                    pValue: dkm_length_format as *const _ as *mut _,
+                    ulValueLen: size_of::<CK_SP800_108_DKM_LENGTH_FORMAT>() as CK_ULONG,
+                },
+                PrfDataParamType::ByteArray(data) => CK_PRF_DATA_PARAM {
+                    type_: CK_SP800_108_BYTE_ARRAY,
+                    pValue: data.as_ptr() as *mut _,
+                    ulValueLen: data
+                        .len()
+                        .try_into()
+                        .expect("length of PRF data parameter does not fit in CK_ULONG"),
+                },
+            },
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Container for information on an additional key to be derived.
+#[derive(Debug)]
+pub struct DerivedKey {
+    template: Pin<Box<[CK_ATTRIBUTE]>>,
+    handle: CK_OBJECT_HANDLE,
+}
+
+impl DerivedKey {
+    /// Construct template for additional key to be derived by KDF.
+    ///
+    /// # Arguments
+    ///
+    /// * `template` - The template for the key to be derived.
+    pub fn new(template: &[Attribute]) -> Self {
+        let template: Box<[CK_ATTRIBUTE]> = template.iter().map(Into::into).collect();
+        let template = Pin::new(template);
+
+        Self {
+            template,
+            handle: CK_INVALID_HANDLE,
+        }
+    }
+
+    /// Return handle for derived key, if it has been created yet
+    pub fn handle(&self) -> Option<ObjectHandle> {
+        if self.handle == CK_INVALID_HANDLE {
+            None
+        } else {
+            Some(ObjectHandle::new(self.handle))
+        }
+    }
+}
+
+impl From<&mut DerivedKey> for CK_DERIVED_KEY {
+    fn from(value: &mut DerivedKey) -> Self {
+        CK_DERIVED_KEY {
+            pTemplate: value.template.as_ptr() as CK_ATTRIBUTE_PTR,
+            ulAttributeCount: value
+                .template
+                .len()
+                .try_into()
+                .expect("number of attributes in template does not fit in CK_ULONG"),
+            phKey: &mut value.handle as CK_OBJECT_HANDLE_PTR,
+        }
+    }
+}
+
+/// NIST SP 800-108 (aka KBKDF) counter and double pipeline-mode parameters.
+///
+/// This structure wraps a `CK_SP800_108_KDF_PARAMS` structure.
+#[derive(Debug)]
+pub struct KbkdfParams<'a> {
+    /// Holds own data so that we have a contiguous memory region to give to backend
+    _additional_derived_keys: Option<Pin<Box<[CK_DERIVED_KEY]>>>,
```

Comment:
Why the underscore?

Reply:
Since we don't use this member, the lint check complains without it. It's just an internal member needed to hold the slice of additional key PKCS#11 structs so that it remains valid for the lifetime of the params object (this is a consequence of what I explained in the Design Justifications section of the PR body)

Reply:
Don't you use though on line 286 when setting its value? If so I think it's ok to remove the underscore.

Reply:
It seems because I'm deriving `Debug`, it complains: https://github.com/rust-lang/rust/issues/88900. I would have to access the field at least once to avoid the compiler warning. I'm guessing keeping the `Debug` trait is important, right?